### ISSUE TITLE
[5.0] The same CSS and JS assets can be loaded multiple times

### ISSUE THREAD

**Issue by pxpm** [Laravel-Backpack/CRUD#2283]
# Bug report

### What i did ?

Created a crud with a select2, select2_multiple, select2_from_ajax, select_2_from_ajax_multiple, select2_nested, select2_grouped  ... etc, basically all fields that require select2.js

### What I expected to happen

Select2.js and select2.css should only be loaded once.

### What happened

Select2.js and select2.css __ARE LOADED ONE TIME PER FIELD TYPE__

As they are all different "field types" but use the same files. 

My proposal would be something along this lines:

Create a config array with asset groups: 

```
'asset_groups' => [
    [
        'select2' => [
            'css' => ['css/select2.css','css/select2_custom.css'],
            'js' => ['js/select2.js','js/select2custom.js'],
        ],
        'datepicker' => [
            'css' => ['css/datepicker.css','css/datepicker.css'],
            'js' => ['js/datepicker.js','js/datepicker.js'],
        ],
        'custom' => [
            'js' => ['js/selects.js'],
        ],
    ]
]
```

In the field we could do something along this lines: 

```
//select2, select2_multiple, select2_from_ajax .. etc

$crud->pushAssets(['select2', 'custom'])

if ($crud->fieldTypeNotLoaded($field)) {
        $crud->markFieldTypeAsLoaded($field);
}

@push('crud_fields_scripts')
foreach($crud->getAssets('js') as $asset) {
<script src="$asset" >
}
@endpush

@push('crud_fields_styles')
foreach($crud->getAssets('css') as $asset) {
<style src="$asset" >
}
@endpush
```

This config file would not be `public`. Otherwise publishing updates to it gonna be an hassle! We can provide ways of developers taking advantage of this functionality for their own fields, maybe let them create a config_assets.php in their config directory and merge with our internal. 

What do you think @tabacitu ?

Best,
Pedro

**Comment by tabacitu** [comment]
This is something that's been bugging me too. It's not a big bug, in my opinion, but it _is_ an annoyance. Not something I would spend time on fixing right now, in 4.0 - I'd postpone it to 4.1.

But it's maybe time to brainstorm solutions for it, so that when we do work on this, it's clear what to do.

One problem that I see with your solution above (let's call it Solution 1) is that I don't see how it would work for big custom ```<script>``` or ```<style>``` blocks. You can't define those in an array. So the LINK and SCRIPT  inclusions would NOT be duplicated, yes, but custom blocks would.

Your solution above is pretty similar in scope to what I had in mind - [loading JS and CSS assets from a PHP array](https://github.com/Laravel-Backpack/CRUD/issues/1887). It's been scrapped, but maybe we should reopen it. [In a different thread](https://github.com/Laravel-Backpack/CRUD/pull/1890#issuecomment-494686085) I've mentioned [orchestral/asset](https://github.com/orchestral/asset) as a possible solution, but the problem with that one is that it doesn't support CDNs, only CSS and JS _files_. I do like their syntax, though. 

The solution I had in mind for this problem is a similar to yours, but not really 😆  Let me know what you think about it below.

For example, for ```fields/select2.blade.php```, instead of doing:

```php
@if ($crud->fieldTypeNotLoaded($field))
    @php
        $crud->markFieldTypeAsLoaded($field);
    @endphp

    {{-- FIELD CSS - will be loaded in the after_styles section --}}
    @push('crud_fields_styles')
        <!-- include select2 css-->
        <link href="{{ asset('packages/select2/dist/css/select2.min.css') }}" rel="stylesheet" type="text/css" />
        <link href="{{ asset('packages/select2-bootstrap-theme/dist/select2-bootstrap.min.css') }}" rel="stylesheet" type="text/css" />
    @endpush

    {{-- FIELD JS - will be loaded in the after_scripts section --}}
    @push('crud_fields_scripts')
        <!-- include select2 js-->
        <script src="{{ asset('packages/select2/dist/js/select2.full.min.js') }}"></script>
        @if (app()->getLocale() !== 'en')
        <script src="{{ asset('packages/select2/dist/js/i18n/' . app()->getLocale() . '.js') }}"></script>
        @endif

        <script>
            function bpFieldInitSelect2Element(element) {
                // element will be a jQuery wrapped DOM node
                if (!element.hasClass("select2-hidden-accessible")) {
                    element.select2({
                        theme: "bootstrap"
                    });
                }
            }
        </script>
    @endpush

@endif
```

# Solution 2

We could have something like:

```php
    {{-- FIELD CSS - will be loaded in the after_styles section --}}
    @push('crud_fields_styles')
        <!-- include select2 css-->
        @loadCssFile(asset('packages/select2/dist/css/select2.min.css'))
        @loadCssFile(asset('packages/select2-bootstrap-theme/dist/select2-bootstrap.min.css'))
    @endpush

    {{-- FIELD JS - will be loaded in the after_scripts section --}}
    @push('crud_fields_scripts')
        <!-- include select2 js-->
        @loadJsFile(asset('packages/select2/dist/js/select2.full.min.js'))
        @if (app()->getLocale() !== 'en')
           @loadJsFile(asset('packages/select2/dist/js/i18n/' . app()->getLocale() . '.js'))
        @endif

        @loadBlock
        <script>
            function bpFieldInitSelect2Element(element) {
                // element will be a jQuery wrapped DOM node
                if (!element.hasClass("select2-hidden-accessible")) {
                    element.select2({
                        theme: "bootstrap"
                    });
                }
            }
        </script>
        @endBlock
    @endpush
```

So basically, instead of echoing the JS and CSS assets using ```<link>``` and ```<script>``` tags, we could be using some Blade directives; those directives store a record of everything they've echoed, and if that thing has already been echoed, it won't echo it again.

This is just my current way of thinking about this, but this _should_ be a good general solution for this, right?

And we could have Blade directives that handle every use case:

```php
// local CSS (runs asset() on the parameter)
@loadLocalCss('path/to/file.css')
// remote CSS (from a CDN)
@loadRemoteCss('http://cdn.smth.com/path/to/file.css')
// inline HTML block
@loadBlock
   <style>
       .red { color: red; }
   </style>
@endBlock

// local JS (runs asset() on the parameter)
@loadLocalJs('path/to/file.js')
// remote JS (from a CDN)
@loadRemoteJs('http://cdn.smth.com/path/to/file.js')
// inline JS
@loadBlock
        <script>
            function bpFieldInitSelect2Element(element) {
                // element will be a jQuery wrapped DOM node
                if (!element.hasClass("select2-hidden-accessible")) {
                    element.select2({
                        theme: "bootstrap"
                    });
                }
            }
        </script>
@endBlock
```

Naming and syntax needs some more work. 

# Solution 3

I guess this solution could be made even more general by making only ONE Blade directive, something like ```@loadOnce```. Which could work:

```php
    {{-- FIELD CSS - will be loaded in the after_styles section --}}
    @push('crud_fields_styles')
        <!-- include select2 css-->
        @loadOnce
        <link href="{{ asset('packages/select2/dist/css/select2.min.css') }}" rel="stylesheet" type="text/css" />
        @endloadOnce
        @loadOnce
        <link href="{{ asset('packages/select2-bootstrap-theme/dist/select2-bootstrap.min.css') }}" rel="stylesheet" type="text/css" />
        @endLoadOnce
    @endpush

    {{-- FIELD JS - will be loaded in the after_scripts section --}}
    @push('crud_fields_scripts')
        <!-- include select2 js-->
        @loadOnce
        <script src="{{ asset('packages/select2/dist/js/select2.full.min.js') }}"></script>
        @endLoadOnce
        @if (app()->getLocale() !== 'en')
        @loadOnce
        <script src="{{ asset('packages/select2/dist/js/i18n/' . app()->getLocale() . '.js') }}"></script>
        @endLoadOnce
        @endif

        @loadOnce
        <script>
            function bpFieldInitSelect2Element(element) {
                // element will be a jQuery wrapped DOM node
                if (!element.hasClass("select2-hidden-accessible")) {
                    element.select2({
                        theme: "bootstrap"
                    });
                }
            }
        </script>
        @endLoadOnce
    @endpush
```

But... it seems like we're littering the code with ```@loadOnce``` statements, since they work best if specified on an individual block, that loads one asset / inline block. So it's simpler, but uglier...

# Solution 4

Maybe a mix of the two solutions above would be better:

```php
@loadOnce('path/to/file.css')->type('link')
@loadOnce('path/to/file.css')->type('link')->local()
@loadOnce('http://cdn.com/path/to/file.css')->type('link')->remote()

@loadOnce
   <script>
     // something here
   </script>
@endLoadOnce

@loadOnce('path/to/file.js')->type('script')
@loadOnce('path/to/file.js')->type('script')->local()
@loadOnce('http://cdn.com/path/to/file.js')->type('script')->remote()

// ideally you wouldn't need to specify link/script or local/remote, it would figure that out itself 
// so the most used syntax would be the shorthand
@loadOnce('path/to/file.js')
```

The end result for the select2 field above would be:

```php
    {{-- FIELD CSS - will be loaded in the after_styles section --}}
    @push('crud_fields_styles')
        <!-- include select2 css-->
        @loadOnce('packages/select2/dist/css/select2.min.css')
        @loadOnce('packages/select2-bootstrap-theme/dist/select2-bootstrap.min.css')
    @endpush

    {{-- FIELD JS - will be loaded in the after_scripts section --}}
    @push('crud_fields_scripts')
        <!-- include select2 js-->
        @loadOnce('packages/select2/dist/js/select2.full.min.js')
        @if (app()->getLocale() !== 'en')
           @loadOnce('packages/select2/dist/js/i18n/' . app()->getLocale() . '.js')
        @endif

        @loadOnce
        <script>
            function bpFieldInitSelect2Element(element) {
                // element will be a jQuery wrapped DOM node
                if (!element.hasClass("select2-hidden-accessible")) {
                    element.select2({
                        theme: "bootstrap"
                    });
                }
            }
        </script>
        @endLoadOnce
    @endpush
```

----

I know this is a lot to take in. But I think it's good to just brainstom every solution we can come up with, and choose the one that would work best 😄 

Let me know what you think. Take your time - this isn't something we will launch until February (4.1), I think.

Cheers!

**Comment by tabacitu** [comment]
I thought we could put together a great solution for this before the 4.1 launch. But it's clear to me that we don't have time to do that. So let's move this to the next release - 6 months from now.

**Comment by juventus18** [comment]
I did some more digging around for my issue today #2655. I think these are related... the select2 styles appear to be overridden in the bundle, but when the select2 fields load their own CSS, those are pushed after the bundle (and any custom files) and thus, the styles are re-overridden by the defaults.  In addition to loading these files only once, we also need to ensure the default styles are loaded *before* any customizations take place (either from Backpack/CoreUI or user custom CSS)

**Comment by tabacitu** [comment]
@juventus18 I don't think that problem is related to this - as far as I know the ```select2``` styles are NOT included in the bundle. The only styles for ```select2``` are the ones loaded by the field.

But since we're on this topic. There's a surprisingly simple solution for this issue, here - https://github.com/Laravel-Backpack/CRUD/pull/2652 . Currently considering merging it to 4.1.

**Comment by juventus18** [comment]
Sorry, you're right... there are no select2 CSS modifications in the bundle, I was looking at `<select>` styling (which ends up getting completely hidden by select2`). At any rate, the styles for <select> in the bundle are what we want to apply to select2. But the problem still stands. Right now, with the way files are currently loading, it is not possible to override the select2 styles in a reasonable way. This is because the select2 styles (both `select2.min.css` and `select2-booststrap.min.css`) are loaded multiple times) after `bundle.css` and any custom css you've included from the base config[styles]. So, even if I try to override the style in a custom .css, since the select2 css is loaded after, that is the styling that will be applied..

in custom.css (or bundle.css)
```
.select2-container--bootstrap .select2-selection--single {
    height: 39px;
}
```
in select2-bootstrap.min.css (loaded after)
```
.select2-container--bootstrap .select2-selection--single {
    height: 34px // <--this style will apply
}

That is the specific issue as this relates to select2, but if I am to summarize it generally for this issue, it is this: 3rd party package styles should be loaded before `bundle.css` and custom styles. So the loading order should be like this:

1. 3rd party packages
2. bundle.css
3. custom css

The solution you're developing here should take this into account.

**Comment by tabacitu** [comment]
True! Will do. Closing this in favour of #2652 . I think it might be the perfect solution for this problem.
